nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py

```diff
@@ -618,17 +618,27 @@ def collate_fn(self, batch):
 
         position_ids: List[List[int]] = []
         cu_seqlens: List[List[int]] = []
+        cu_seqlens_unpadded: List[List[int]] = []
         for item in batch:
             position_ids.append([])
             cu_seqlens.append([0])
+            cu_seqlens_unpadded.append([0])
             seqlens = np.array(item['seq_boundaries'][1:]) - np.array(item['seq_boundaries'][:-1])
             for l in seqlens:
                 # length minus 1 because input_ids is truncated by 1 for labels
                 position_ids[-1].extend(list(range(l - 1)))
                 cu_seqlens[-1].append(cu_seqlens[-1][-1] + l - 1)
+
             # set last seq to the max seq len because rope and attn kernels expect no padding
             cu_seqlens[-1][-1] = max_length
 
+            for i in range(len(item['seq_boundaries']) - 1):
+                seqlen_unpadded = np.argmin(
+                    np.array(item['input_ids'][item['seq_boundaries'][i] : item['seq_boundaries'][i + 1] - 1])
+                )
```

Comment:
We assume pad_id is the minimum value in `input_ids`? is this always true?

Reply:
I used the dataset.tokenizer.eos_id as pad_id and it's the minimum value in input_ids.